framework/src/bcs/ArrayLowerDIntegratedBC.C

```diff
@@ -0,0 +1,197 @@
+  : ArrayIntegratedBC(parameters),
+    _lowerd_var(*getArrayVar("lowerd_variable", 0)),
+    _lambda(_is_implicit ? _lowerd_var.slnLower() : _lowerd_var.slnLowerOld()),
+
+    _phi_lambda(_lowerd_var.phiLower()),
+    _test_lambda(_lowerd_var.phiLower())
+{
+  if (_lowerd_var.activeSubdomains().count(Moose::BOUNDARY_SIDE_LOWERD_ID) == 0)
+    mooseError("'lowerd_variable' must be defined on BOUNDARY_SIDE_LOWERD_SUBDOMAIN subdomain");
+
+  if (_lowerd_var.count() != _count)
+    mooseError("The number of components of 'lowerd_variable' must be equal to the number of "
+               "components of 'variable'");
+}
+
+void
+ArrayLowerDIntegratedBC::computeResidual()
+{
+  ArrayIntegratedBC::computeResidual();
+
+  prepareVectorTagLower(_assembly, _lowerd_var.number());
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    initLowerDQpResidual();
+    for (_i = 0; _i < _test_lambda.size(); _i++)
+    {
+      _work_vector.setZero();
+      computeLowerDQpResidual(_work_vector);
+      mooseAssert(_work_vector.size() == _count,
+                  "Size of local residual is not equal to the number of array variable compoments");
+      _work_vector *= _JxW[_qp] * _coord[_qp];
+      _assembly.saveLocalArrayResidual(_local_re, _i, _test_lambda.size(), _work_vector);
+    }
+  }
+
+  accumulateTaggedLocalResidual();
+}
+
+void
+ArrayLowerDIntegratedBC::computeJacobian()
+{
+  ArrayIntegratedBC::computeJacobian();
+
+  computeLowerDJacobian(Moose::LowerLower);
+  computeLowerDJacobian(Moose::LowerPrimary);
+  computeLowerDJacobian(Moose::PrimaryLower);
+}
+
+void
+ArrayLowerDIntegratedBC::computeLowerDJacobian(Moose::ConstraintJacobianType type)
+{
+  mooseAssert(type == Moose::LowerLower || type == Moose::LowerPrimary ||
+                  type == Moose::PrimaryLower,
+              "Jacobian types must have lower in computeLowerDJacobian");
+
+  const ArrayVariableTestValue & test_space =
+      (type == Moose::LowerLower || type == Moose::LowerPrimary) ? _test_lambda : _test;
+  unsigned int ivar = (type == Moose::LowerLower || type == Moose::LowerPrimary)
+                          ? _lowerd_var.number()
+                          : _var.number();
+
+  const ArrayVariableTestValue & loc_phi =
+      (type == Moose::LowerLower || type == Moose::PrimaryLower) ? _phi_lambda : _phi;
+  unsigned int jvar = (type == Moose::LowerLower || type == Moose::PrimaryLower)
+                          ? _lowerd_var.number()
+                          : _var.number();
+
+  prepareMatrixTagLower(_assembly, ivar, jvar, type);
+
+  if (_local_ke.n() == 0 || _local_ke.m() == 0)
+    return;
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    initLowerDQpJacobian(type);
+    for (_i = 0; _i < test_space.size(); _i++)
+      for (_j = 0; _j < loc_phi.size(); _j++)
+      {
+        RealEigenVector v = _JxW[_qp] * _coord[_qp] * computeLowerDQpJacobian(type);
+        _assembly.saveDiagLocalArrayJacobian(
```

Comment:
difficult because we do not have something like `ArrayResidualObject` to put this function. I will add some more comments in `Assembly.h` about this function. We can think of a better way later.

Reply:
Do these operations in an `ArrayResidualObject` will make more sense . We may create one?

Reply:
It might have to be in an interface. It could be an overkill.